Python/ceval_gil.c

```diff
@@ -57,113 +57,60 @@
 #define _Py_atomic_load_relaxed_int32(ATOMIC_VAL) _Py_atomic_load_relaxed(ATOMIC_VAL)
 #endif
 
-/* This can set eval_breaker to 0 even though gil_drop_request became
-   1.  We believe this is all right because the eval loop will release
-   the GIL eventually anyway. */
+
 static inline void
-COMPUTE_EVAL_BREAKER(PyInterpreterState *interp,
-                     struct _ceval_runtime_state *ceval,
-                     struct _ceval_state *ceval2)
+update_eval_breaker_from_thread(PyInterpreterState *interp, PyThreadState *tstate)
 {
-    _Py_atomic_store_relaxed(&ceval2->eval_breaker,
-        _Py_atomic_load_relaxed_int32(&ceval2->gil_drop_request)
-        | (_Py_atomic_load_relaxed_int32(&ceval->signals_pending)
-           && _Py_ThreadCanHandleSignals(interp))
-        | (_Py_atomic_load_relaxed_int32(&ceval2->pending.calls_to_do))
-        | (_Py_IsMainThread() && _Py_IsMainInterpreter(interp)
-           &&_Py_atomic_load_relaxed_int32(&ceval->pending_mainthread.calls_to_do))
-        | ceval2->pending.async_exc
-        | _Py_atomic_load_relaxed_int32(&ceval2->gc_scheduled));
+    if (tstate == NULL) {
+        return;
+    }
+    int32_t calls_to_do;
+    if (_Py_ThreadCanHandleSignals(interp)) {
+        calls_to_do = _Py_atomic_load_int32_relaxed(
+            &_PyRuntime.ceval.pending_mainthread.calls_to_do);
+        if (_Py_atomic_load(&_PyRuntime.signals.is_tripped)) {
+            _Py_set_eval_breaker_bit(interp, _PY_SIGNALS_PENDING_BIT, 1);
+        }
+    }
+    else {
+        calls_to_do = _Py_atomic_load_int32_relaxed(
+            &interp->ceval.pending.calls_to_do);
+    }
+    if (calls_to_do) {
+        _Py_set_eval_breaker_bit(interp, _PY_CALLS_TO_DO_BIT, 1);
+    }
+    if (tstate->async_exc != NULL) {
+        _Py_set_eval_breaker_bit(interp, _PY_ASYNC_EXCEPTION_BIT, 1);
+    }
```

Comment:
Where are we not already setting the bit when `tstate->async_exc` is set, such that we must do so here?

Reply:
If `tstate->async_exc` is set on a sleeping thread, then we need to set the bit when that thread gets the GIL.

Reply:
Let's see if I understand things correctly.  A thread will only ever acquire the GIL either by the eval loop itself or via the C-API:

* C-API[^1]
   * before the eval loop has started
   * while the eval loop is running
   * after the eval loop has finished
* running eval loop--for some bytecode instructions[^2]

On the other hand, all the "pending" actions[^3] happen only are performed in a running eval loop, driven by the eval breaker[^2].  The eval loop calls `_Py_HandlePending()`, which executes in this order:

1. handle signals, if any (**main thread only**)
2. make per-interpreter pending calls
3. make global pending calls (**main thread only**)
4. run GC[^4]
5. release & re-acquire the GIL (if another thread has asked for it)
6. apply the async exception, if applicable (**sort of thread-specific**)

When the GIL is acquired via the C-API (outside a running eval loop), none of that will happen until at least one eval loop is running and it reaches one of the instructions that checks the eval breaker[^2].

When the GIL is acquired by a running eval loop, the first 4 pending events have already just been handled, and the last one (async exception) is about to be.

[^1]: e.g. `PyEval_RestoreThread()` (after earlier releasing it), or for the first time when an interpreter is created or with `threading.Thread.start()`.
[^2]: the small set of instructions that invoke `CHECK_EVAL_BREAKER()`: the intermediate end of loops; after calls; sometimes for `RESUME`; `_JUMP_TO_TOP` and `ENTER_EXECUTOR` (whatever those are for).
[^3]: which are driven by the events represented in this PR
[^4]: GC is also triggered in other places, like the signal-related C-API

----

With all that in mind, the question is: how does the effect of `PyThreadState_SetAsyncExc()` relate to acquiring the GIL?  Hence, why should we worry about `_PY_ASYNC_EXCEPTION_BIT` in `update_eval_breaker_from_thread()`?  `PyThreadState_SetAsyncExc()` always sets `tstate->async_exc` and then sets the eval breaker for the target interpreter.  That eval breaker bit never gets cleared until the targeted thread takes care of event.  So why do we need to set the bit when taking the GIL, regardless of if the thread is sleeping or not?  It seems unnecessary.

That does lead to another question:  why have the async exception be part of the eval breaker at all?  It only only applies to the target thread, but the eval breaker bit is set it will interrupt all threads.  Would it make sense to have a per-thread eval breaker?  Then again, it only matters if async exceptions are used by the community often enough.

Reply:
There's a similar question for signals and the main thread pending calls,.  However, unlike async exceptions, we know that signals are somewhat prevalent.  Maybe a separate per-thread eval breaker would be worth it?  Perhaps it would help to have a separate (generated) variant of the eval loop that also checks a per-thread eval breaker (or checks it exclusively)?

Reply:
FWIW, I don't think we have much room to change the order of actions in `_Py_HandlePending()`.